BLOCKCHAIN APPLICATIONS


1.Convert given string to hash form.
2.To convert encoded value of string to acceptable form we convert to hexadecimal form.

Using hash algorithms sha256:

In [ ]:
import hashlib

str = "neha"


# then sending to sha256()
result = hashlib.sha256(str.encode())
print(result)
# printing the equivalent hexadecimal value.
print("The hexadecimal equivalent of hash is : ", end ="")
res=result.hexdigest()
print(res)
print("Available hash algorithms are ",hashlib.algorithms_available)

<sha256 HASH object @ 0x7eff619425d0>
The hexadecimal equivalent of hash is : a86d61e4c4516ca829dfb5a3b31186e0b2c11268b2d2a225384df86f71b72944
Available hash algorithms are  {'sha3_384', 'sha1', 'blake2s', 'sha384', 'sha3_512', 'md5', 'blake2b', 'shake_128', 'sha3_256', 'shake_256', 'sha3_224', 'sha512', 'sha256', 'sha224'}


Using md5 hash function:

In [ ]:
import hashlib

str = "neha"


# then sending to md5()
result = hashlib.md5(str.encode())
print(result)
# printing the equivalent hexadecimal value.
print("The hexadecimal equivalent of hash is : ", end ="")
res=result.hexdigest()
print(res)
print("Available hash algorithms are ",hashlib.algorithms_available)

<md5 HASH object @ 0x7eff619420c0>
The hexadecimal equivalent of hash is : 262f5bdd0af9098e7443ab1f8e435290
Available hash algorithms are  {'sha3_384', 'sha1', 'blake2s', 'sha384', 'sha3_512', 'md5', 'blake2b', 'shake_128', 'sha3_256', 'shake_256', 'sha3_224', 'sha512', 'sha256', 'sha224'}


Generates encrypted code and create password

In [ ]:
import uuid
import hashlib

import timeit

start = timeit.default_timer()

def hash_password(password):
    # uuid is used to generate a random number
    random_number = uuid.uuid4().hex
    return hashlib.sha256(random_number.encode() + password.encode()).hexdigest() + ':' + random_number

def check_password(hashed_password, user_password):
    password, random_number = hashed_password.split(':')
    return password == hashlib.sha256(random_number.encode() + user_password.encode()).hexdigest()

new_pass = input('Please enter a password: ')
hashed_password = hash_password(new_pass)
print('The string to store in the database(harddrive) is: ' + hashed_password)
old_pass = input('Now please enter the password again to check: ')
if check_password(hashed_password, old_pass):
    print('You entered the right password')
else:
    print('I am sorry but the hash of the password does not match')

stop = timeit.default_timer()

print('Time: ', stop - start)

Please enter a password: neha
The string to store in the database(harddrive) is: a82963db501f1dda59e161dcd41de3b8dc133d1278f3aae5e94742f6f76bd73e:919d250df11d4fedb4fc62a2fbd013d2
Now please enter the password again to check: neha
You entered the right password
Time:  6.463478660999954


In [ ]:
import uuid
import hashlib

import timeit

start = timeit.default_timer()

def hash_password(password):
    # uuid is used to generate a random number
    random_number = uuid.uuid4().hex
    return hashlib.sha256(random_number.encode() + password.encode()).hexdigest() + ':' + random_number

def check_password(hashed_password, user_password):
    password, random_number = hashed_password.split(':')
    return password == hashlib.sha256(random_number.encode() + user_password.encode()).hexdigest()

new_pass = input('Please enter a password: ')
hashed_password = hash_password(new_pass)
print('The string to store in the database(harddrive) is: ' + hashed_password)
old_pass = input('Now please enter the password again to check: ')
if check_password(hashed_password, old_pass):
    print('You entered the right password')
else:
    print('I am sorry but the hash of the password does not match')

stop = timeit.default_timer()

print('Time: ', stop - start)

Please enter a password: lavanya
The string to store in the database(harddrive) is: 9949389f7c5b04b46f249d35de1e079a775cd2f5978090706959663ef09ded38:913340d189874112bfb10c5390af0de7
Now please enter the password again to check: disha
I am sorry but the hash of the password does not match
Time:  24.299284582000155


Generating 20 blocks, concatenating data of current block with prev block and printing encrypted hexadecimal code

In [ ]:
import hashlib
block = [{"data":"22","prevhash":"0"}]
for i in range(20):
    print(i)
    concat=block[i]["data"] + block[i]["prevhash"]
    print("concatenated string ",concat,"encoded string:",concat.encode())
    calhash=hashlib.sha256(concat.encode())
    print(calhash.hexdigest())
    block+=[{"data":"23","prevhash":calhash.hexdigest()}]

0
concatenated string  220 encoded string: b'220'
36790ecd55c2030dc553685bef719df653f413a20cdad1bfd1dc934c76686ddd
1
concatenated string  2336790ecd55c2030dc553685bef719df653f413a20cdad1bfd1dc934c76686ddd encoded string: b'2336790ecd55c2030dc553685bef719df653f413a20cdad1bfd1dc934c76686ddd'
df65d366277e62cd233c7fb2dffd1d2275c32f589e99b0acc6f374e390c81e4e
2
concatenated string  23df65d366277e62cd233c7fb2dffd1d2275c32f589e99b0acc6f374e390c81e4e encoded string: b'23df65d366277e62cd233c7fb2dffd1d2275c32f589e99b0acc6f374e390c81e4e'
acc86ab25c20023a6dc76afc1873528f4d7ad873592a1d77ab907d4426985a88
3
concatenated string  23acc86ab25c20023a6dc76afc1873528f4d7ad873592a1d77ab907d4426985a88 encoded string: b'23acc86ab25c20023a6dc76afc1873528f4d7ad873592a1d77ab907d4426985a88'
abf6f1942501ed9aabbe132cb986b5af5551e70b2938cb7f2d2e1e0eff47eaef
4
concatenated string  23abf6f1942501ed9aabbe132cb986b5af5551e70b2938cb7f2d2e1e0eff47eaef encoded string: b'23abf6f1942501ed9aabbe132cb986b5af5551e70b2938cb7f2d2e

first iteration: 22+0 = 220
from next iteration: 23+ prev hash code
concatenating

In [ ]:
import hashlib
string="0413e2bfe73e0a2df0c275f203399370c7c857f60b88fe16db0d8e212a303c4"
block = [{"data":"23","prevhash":"0"}]

for i in range(100000):
    print(i)

    concat=block[i]["data"] + block[i]["prevhash"]
    #print("concatenated string ",concat,"encoded string:",concat.encode())
    calhash=hashlib.sha256(concat.encode())
    string=calhash.hexdigest()
    print(string)
    print(string[0]+string[1])
    if (string[0]+string[1])=="00":
        break
    block+=[{"data":"23","prevhash":calhash.hexdigest()}]

0
a0eaec5a55dc2f5b2ba523018adc485ff620b9d83509b9f37186a7716e438d21
a0
1
874d17ad5d60ce4cb176a8309576eb38d8ee2e996b91fc25120daa8c80320a57
87
2
5c7aedfb7313bbbbca4d3187b91d03fdd7f2a59afdefb3558e83bd495b4ef980
5c
3
839a0fe708dacf3982ddcfa75ab04101a2431937c90c6f8db5246497d87482fa
83
4
62ae015aa8cbc65874c8187a7d520a36d84d5d98408b37369f325fdc83525417
62
5
6cf7c0a4a08a391eceeaac2c58a3e4e59048331efa7576716ad6af3dc4e6d93c
6c
6
0e77562e517a8a07f06cecbd3952bc7a3468bf1a5f25c9b0c4cfbf67fb7c7a83
0e
7
5adec8b7c664b595518d46bcd2c3620c211ae433ef727662fdc59c07fd47a0ff
5a
8
9a4fafdee1d4aa0673b85a67021d6f9f52ace4468fb6176a947b21c8adfef848
9a
9
a0a4cd124d18d9979eb3618f475462d687fb81c054ba5217afa2c6ada0358bd6
a0
10
9fa8341c147632b6123059b670c0513c668ddd6944bab12edadd92081f751bce
9f
11
200c9a846adf95c4da5be07f9411d78585198a3c036262127fe78504cfbf6faf
20
12
6a1f6f21e92c271eb6a4b8e84c12f69bf9af22d151d51f5fd15089a38276cc5c
6a
13
d81833b3ad9dd055e9dcb6fb30f35dcb6d9b6764121c6bb6c65efd118ea89e9d
d8
14
e9ea03999635f

Terminates when code reaches 00

In [ ]:

import hashlib
import timeit
start = timeit.default_timer()
string="0413e2bfe73e0a2df0c275f203399370c7c857f60b88fe16db0d8e212a303c4"
#print(string[0]+string[1])

block = [{"data":"23","prevhash":"0413e2bfe73e0a2df0c275f203399370c7c857f60b88fe16db0d8e212a303c4"}]

for i in range(10000):
    print(i)

    concat=block[0]["data"] + block[0]["prevhash"] + str(i)
    #print("concatenated string ",concat,"encoded string:",concat.encode())
    calhash=hashlib.sha256(concat.encode())
    string=calhash.hexdigest()

print(string)
print(string[0]+string[1]+string[2])
    if (string[0]+string[1]+string[2])=="000":
        break
stop = timeit.default_timer()

print('Time: ', stop - start)

In [ ]:
import hashlib
array=["1","2","3","4","5","6","7","8","1","2","3","4","5","6","7","8"]
#print(len(array))
arr1=[]
array1=[]
def merkle(array1):
	outputarray=[]
	i=0
	l=len(array1)/2
	l=int(l)
	for j in range(l):
		#print(i,i+1)
		concat = array1[i]+array1[i+1]
		hash=hashlib.sha256(concat.encode()).hexdigest()
		outputarray+=[hash]
		i=i+2
	print("hash array length ",len(outputarray),"\n",outputarray,"\n")
	return outputarray
while len(array)!=1:
	outputarray=merkle(array)
	#print("length of final array",len(outputarray))
	array=outputarray

hash array length  8 
 ['6b51d431df5d7f141cbececcf79edf3dd861c3b4069f0b11661a3eefacbba918', '86e50149658661312a9e0b35558d84f6c6d3da797f552a9657fe0558ca40cdef', '7688b6ef52555962d008fff894223582c484517cea7da49ee67800adc7fc8866', '349c41201b62db851192665c504b350ff98c6b45fb62a8a2161f78b6534d8de9', '6b51d431df5d7f141cbececcf79edf3dd861c3b4069f0b11661a3eefacbba918', '86e50149658661312a9e0b35558d84f6c6d3da797f552a9657fe0558ca40cdef', '7688b6ef52555962d008fff894223582c484517cea7da49ee67800adc7fc8866', '349c41201b62db851192665c504b350ff98c6b45fb62a8a2161f78b6534d8de9'] 

hash array length  4 
 ['88cd668c2056e926cf9f6dad3acbeebf0c1e093da5ab7aceb244e65661d7e35e', 'fdefe3e799c1f869de615c07e99b7969023909c30fa162eb2a4d69d93ac74ffb', '88cd668c2056e926cf9f6dad3acbeebf0c1e093da5ab7aceb244e65661d7e35e', 'fdefe3e799c1f869de615c07e99b7969023909c30fa162eb2a4d69d93ac74ffb'] 

hash array length  2 
 ['5cd190530acea64c32e5c777d3ab607866be5012dfb34cb2c5aa7f3d89dca7c5', '5cd190530acea64c32e5c777d3ab607866be501

length/2: 8,4,2,1

A nonce is an abbreviation for "number only used once," which is a number added to a hashed—or encrypted—block in a blockchain that, when rehashed, meets the difficulty level restrictions. The nonce is the number that blockchain miners are solving for, in order to receive cryptocurrency.

TRANSACTION IS BETWEEN TWO PUBLIC KEYS...
Sender encrypts and reciever decrypts using pvt key

In [ ]:
const Block = require("./Block.js");

class Blockchain {
  constructor () {
    this.blockchain = [Block.genesis()];
  }

  get() {
    return this.blockchain;
  }

  get latestBlock() {
    return this.blockchain[this.blockchain.length - 1];
  }
};

module.exports = Blockchain;


SyntaxError: ignored

In [ ]:


pragma solidity ^0.8.0;

/**
 * @dev Interface of the ERC20 standard as defined in the EIP. Does not include
 * the optional functions; to access them see `ERC20Detailed`.
 */
interface IERC20 {
    /**
     * @dev Returns the amount of tokens in existence.
     */
    function totalSupply() external view returns (uint256);

    /**
     * @dev Returns the amount of tokens owned by `account`.
     */
    function balanceOf(address account) external view returns (uint256);

    /**
     * @dev Moves `amount` tokens from the caller's account to `recipient`.
     *
     * Returns a boolean value indicating whether the operation succeeded.
     *
     * Emits a `Transfer` event.
     */
    function transfer(address recipient, uint256 amount)
        external
        returns (bool);

    /**
     * @dev Returns the remaining number of tokens that `spender` will be
     * allowed to spend on behalf of `owner` through `transferFrom`. This is
     * zero by default.
     *
     * This value changes when `approve` or `transferFrom` are called.
     */
    function allowance(address owner, address spender)
        external
        view
        returns (uint256);

    /**
     * @dev Sets `amount` as the allowance of `spender` over the caller's tokens.
     *
     * Returns a boolean value indicating whether the operation succeeded.
     *
     * > Beware that changing an allowance with this method brings the risk
     * that someone may use both the old and the new allowance by unfortunate
     * transaction ordering. One possible solution to mitigate this race
     * condition is to first reduce the spender's allowance to 0 and set the
     * desired value afterwards:
     * https://github.com/ethereum/EIPs/issues/20#issuecomment-263524729
     *
     * Emits an `Approval` event.
     */
    function approve(address spender, uint256 amount) external returns (bool);

    /**
     * @dev Moves `amount` tokens from `sender` to `recipient` using the
     * allowance mechanism. `amount` is then deducted from the caller's
     * allowance.
     *
     * Returns a boolean value indicating whether the operation succeeded.
     *
     * Emits a `Transfer` event.
     */
    function transferFrom(
        address sender,
        address recipient,
        uint256 amount
    ) external returns (bool);

    /**
     * @dev Emitted when `value` tokens are moved from one account (`from`) to
     * another (`to`).
     *
     * Note that `value` may be zero.
     */
    event Transfer(address indexed from, address indexed to, uint256 value);

    /**
     * @dev Emitted when the allowance of a `spender` for an `owner` is set by
     * a call to `approve`. `value` is the new allowance.
     */
    event Approval(
        address indexed owner,
        address indexed spender,
        uint256 value
    );
}

// File: contracts\open-zeppelin-contracts\math\SafeMath.sol

pragma solidity ^0.8.0;

/**
 * @dev Wrappers over Solidity's arithmetic operations with added overflow
 * checks.
 *
 * Arithmetic operations in Solidity wrap on overflow. This can easily result
 * in bugs, because programmers usually assume that an overflow raises an
 * error, which is the standard behavior in high level programming languages.
 * `SafeMath` restores this intuition by reverting the transaction when an
 * operation overflows.
 *
 * Using this library instead of the unchecked operations eliminates an entire
 * class of bugs, so it's recommended to use it always.
 */
library SafeMath {
    /**
     * @dev Returns the addition of two unsigned integers, reverting on
     * overflow.
     *
     * Counterpart to Solidity's `+` operator.
     *
     * Requirements:
     * - Addition cannot overflow.
     */
    function add(uint256 a, uint256 b) internal pure returns (uint256) {
        uint256 c = a + b;
        require(c >= a, "SafeMath: addition overflow");

        return c;
    }

    /**
     * @dev Returns the subtraction of two unsigned integers, reverting on
     * overflow (when the result is negative).
     *
     * Counterpart to Solidity's `-` operator.
     *
     * Requirements:
     * - Subtraction cannot overflow.
     */
    function sub(uint256 a, uint256 b) internal pure returns (uint256) {
        require(b <= a, "SafeMath: subtraction overflow");
        uint256 c = a - b;

        return c;
    }

    /**
     * @dev Returns the multiplication of two unsigned integers, reverting on
     * overflow.
     *
     * Counterpart to Solidity's `*` operator.
     *
     * Requirements:
     * - Multiplication cannot overflow.
     */
    function mul(uint256 a, uint256 b) internal pure returns (uint256) {
        // Gas optimization: this is cheaper than requiring 'a' not being zero, but the
        // benefit is lost if 'b' is also tested.
        // See: https://github.com/OpenZeppelin/openzeppelin-solidity/pull/522
        if (a == 0) {
            return 0;
        }

        uint256 c = a * b;
        require(c / a == b, "SafeMath: multiplication overflow");

        return c;
    }

    /**
     * @dev Returns the integer division of two unsigned integers. Reverts on
     * division by zero. The result is rounded towards zero.
     *
     * Counterpart to Solidity's `/` operator. Note: this function uses a
     * `revert` opcode (which leaves remaining gas untouched) while Solidity
     * uses an invalid opcode to revert (consuming all remaining gas).
     *
     * Requirements:
     * - The divisor cannot be zero.
     */
    function div(uint256 a, uint256 b) internal pure returns (uint256) {
        // Solidity only automatically asserts when dividing by 0
        require(b > 0, "SafeMath: division by zero");
        uint256 c = a / b;
        // assert(a == b * c + a % b); // There is no case in which this doesn't hold

        return c;
    }

    /**
     * @dev Returns the remainder of dividing two unsigned integers. (unsigned integer modulo),
     * Reverts when dividing by zero.
     *
     * Counterpart to Solidity's `%` operator. This function uses a `revert`
     * opcode (which leaves remaining gas untouched) while Solidity uses an
     * invalid opcode to revert (consuming all remaining gas).
     **/
     * Requirements:
     * - The divisor cannot be zero.
     */
    function mod(uint256 a, uint256 b) internal pure returns (uint256) {
        require(b != 0, "SafeMath: modulo by zero");
        return a % b;
    }
}

// File: contracts\open-zeppelin-contracts\token\ERC20\ERC20.sol

pragma solidity ^0.8.0;

/**
 * @dev Implementation of the `IERC20` interface.
 *
 * This implementation is agnostic to the way tokens are created. This means
 * that a supply mechanism has to be added in a derived contract using `_mint`.
 * For a generic mechanism see `ERC20Mintable`.
 *
 * *For a detailed writeup see our guide [How to implement supply
 * mechanisms](https://forum.zeppelin.solutions/t/how-to-implement-erc20-supply-mechanisms/226).*
 *
 * We have followed general OpenZeppelin guidelines: functions revert instead
 * of returning `false` on failure. This behavior is nonetheless conventional
 * and does not conflict with the expectations of ERC20 applications.
 *
 * Additionally, an `Approval` event is emitted on calls to `transferFrom`.
 * This allows applications to reconstruct the allowance for all accounts just
 * by listening to said events. Other implementations of the EIP may not emit
 * these events, as it isn't required by the specification.
 *
 * Finally, the non-standard `decreaseAllowance` and `increaseAllowance`
 * functions have been added to mitigate the well-known issues around setting
 * allowances. See `IERC20.approve`.
 */
contract Token is IERC20 {
    using SafeMath for uint256;

    mapping(address => uint256) private _balances;

    mapping(address => mapping(address => uint256)) private _allowances;

    uint256 private _totalSupply;

    address public tokenOwner;

    uint256 public whaleAmount;

    bool public antiWhale;

    /**
     * @dev See `IERC20.totalSupply`.
     */
    function totalSupply() public override view returns (uint256) {
        return _totalSupply;
    }

    /**
     * @dev See `IERC20.balanceOf`.
     */
    function balanceOf(address account) public override view returns (uint256) {
        return _balances[account];
    }

    /**
     * @dev See `IERC20.transfer`.
     *
     * Requirements:
     *
     * - `recipient` cannot be the zero address.
     * - the caller must have a balance of at least `amount`.
     */
    function transfer(address recipient, uint256 amount) public override returns (bool) {
        _transfer(msg.sender, recipient, amount);
        return true;
    }

    /**
     * @dev See `IERC20.allowance`.
     */
    function allowance(address owner, address spender)
        public override
        view
        returns (uint256)
    {
        return _allowances[owner][spender];
    }

    /**
     * @dev See `IERC20.approve`.
     *
     * Requirements:
     *
     * - `spender` cannot be the zero address.
     */
    function approve(address spender, uint256 value) public override returns (bool) {
        _approve(msg.sender, spender, value);
        return true;
    }

    /**
     * @dev See `IERC20.transferFrom`.
     *
     * Emits an `Approval` event indicating the updated allowance. This is not
     * required by the EIP. See the note at the beginning of `ERC20`;
     *
     * Requirements:
     * - `sender` and `recipient` cannot be the zero address.
     * - `sender` must have a balance of at least `value`.
     * - the caller must have allowance for `sender`'s tokens of at least
     * `amount`.
     */
    function transferFrom(
        address sender,
        address recipient,
        uint256 amount
    ) public override returns (bool) {
        _transfer(sender, recipient, amount);
        _approve(
            sender,
            msg.sender,
            _allowances[sender][msg.sender].sub(amount)
        );
        return true;
    }

    /**
     * @dev Atomically increases the allowance granted to `spender` by the caller.
     *
     * This is an alternative to `approve` that can be used as a mitigation for
     * problems described in `IERC20.approve`.
     *
     * Emits an `Approval` event indicating the updated allowance.
     *
     * Requirements:
     *
     * - `spender` cannot be the zero address.
     */
    function increaseAllowance(address spender, uint256 addedValue)
        public
        returns (bool)
    {
        _approve(
            msg.sender,
            spender,
            _allowances[msg.sender][spender].add(addedValue)
        );
        return true;
    }

    /**
     * @dev Atomically decreases the allowance granted to `spender` by the caller.
     *
     * This is an alternative to `approve` that can be used as a mitigation for
     * problems described in `IERC20.approve`.
     *
     * Emits an `Approval` event indicating the updated allowance.
     *
     * Requirements:
     *
     * - `spender` cannot be the zero address.
     * - `spender` must have allowance for the caller of at least
     * `subtractedValue`.
     */
    function decreaseAllowance(address spender, uint256 subtractedValue)
        public
        returns (bool)
    {
        _approve(
            msg.sender,
            spender,
            _allowances[msg.sender][spender].sub(subtractedValue)
        );
        return true;
    }

    /**
     * @dev Moves tokens `amount` from `sender` to `recipient`.
     *
     * This is internal function is equivalent to `transfer`, and can be used to
     * e.g. implement automatic token fees, slashing mechanisms, etc.
     *
     * Emits a `Transfer` event.
     *
     * Requirements:
     *
     * - `sender` cannot be the zero address.
     * - `recipient` cannot be the zero address.
     * - `sender` must have a balance of at least `amount`.
     */
    function _transfer(
        address sender,
        address recipient,
        uint256 amount
    ) internal {
        require(sender != address(0), "ERC20: transfer from the zero address");
        require(recipient != address(0), "ERC20: transfer to the zero address");

        if (antiWhale) {
            require(amount <= whaleAmount, "Exceeded whale amount");
        }

        _balances[sender] = _balances[sender].sub(amount);
        _balances[recipient] = _balances[recipient].add(amount);
        emit Transfer(sender, recipient, amount);
    }*/

    /** @dev Creates `amount` tokens and assigns them to `account`, increasing
     * the total supply.
     *
     * Emits a `Transfer` event with `from` set to the zero address.
     *
     * Requirements
     *
     * - `to` cannot be the zero address.
     */
    function _mint(address account, uint256 amount) internal {
        require(account != address(0), "ERC20: mint to the zero address");

        _totalSupply = _totalSupply.add(amount);
        _balances[account] = _balances[account].add(amount);
        emit Transfer(address(0), account, amount);
    }

    /**
     * @dev Destroys `amount` tokens from `account`, reducing the
     * total supply.
     *
     * Emits a `Transfer` event with `to` set to the zero address.
     *
     * Requirements
     *
     * - `account` cannot be the zero address.
     * - `account` must have at least `amount` tokens.
     */
    function _burn(address account, uint256 value) internal {
        require(account != address(0), "ERC20: burn from the zero address");

        _totalSupply = _totalSupply.sub(value);
        _balances[account] = _balances[account].sub(value);
        emit Transfer(account, address(0), value);
    }

    /**
     * @dev Sets `amount` as the allowance of `spender` over the `owner`s tokens.
     *
     * This is internal function is equivalent to `approve`, and can be used to
     * e.g. set automatic allowances for certain subsystems, etc.
     *
     * Emits an `Approval` event.
     *
     * Requirements:
     *
     * - `owner` cannot be the zero address.
     * - `spender` cannot be the zero address.
     */
    function _approve(
        address owner,
        address spender,
        uint256 value
    ) internal {
        require(owner != address(0), "ERC20: approve from the zero address");
        require(spender != address(0), "ERC20: approve to the zero address");

        _allowances[owner][spender] = value;
        emit Approval(owner, spender, value);
    }

    /**
     * @dev Destroys `amount` tokens from `account`.`amount` is then deducted
     * from the caller's allowance.
     **/
     * See `_burn` and `_approve`.
     */
    function _burnFrom(address account, uint256 amount) internal {
        _burn(account, amount);
        _approve(
            account,
            msg.sender,
            _allowances[account][msg.sender].sub(amount)
        );
    }
}

// File: contracts\ERC20\Test.sol

pragma solidity ^0.8.0;

/**
 * @title Test
 * @author Test
 *
 * @dev Standard ERC20 token with burning and optional functions implemented.
 * For full specification of ERC-20 standard see:
 * https://github.com/ethereum/EIPs/blob/master/EIPS/eip-20.md
 */
contract Token is ERC20 {
    string private _name;
    string private _symbol;
    uint8 private _decimals;

	event whaleAmountUpdated(uint256 oldAmount, uint256 newAmount, uint256 time);
	event antiWhaleUpdated(bool status, uint256 time);
    event tokenOwnerUpdated(address owner, address newOwner);


    /**
     * @dev Constructor.
     * @param name_ name of the token
     * @param symbol_ symbol of the token, 3-4 chars is recommended
     * @param decimals_ number of decimal places of one token unit, 18 is widely used
     * @param totalSupply_ total supply of tokens in lowest units (depending on decimals)
     * @param _antiWhale to enable the antiwhale feature on/off, by default value is false.
     * @param _whaleAmount whale amount of tokens in lowest units (depending on decimals)
     * @param tokenOwnerAddress address that gets 100% of token supply
     */
    constructor(
        string memory name_,
        string memory symbol_,
        uint8 decimals_,
        uint256 totalSupply_,
        bool _antiWhale,
        uint256 _whaleAmount,
        address tokenOwnerAddress
    ) payable {
    	require(tokenOwnerAddress != address(0), "Owner can't be zero address");
        if(_antiWhale) {
            require(_whaleAmount >= 1000 * (10 ** decimals_), "Whale amount less than 1000");
        }
        _name = name_;
        _symbol = symbol_;
        _decimals = decimals_;
        tokenOwner = tokenOwnerAddress;
        whaleAmount = _whaleAmount;
        antiWhale = _antiWhale;

        // set tokenOwnerAddress as owner of all tokens and the owner has the control of antiWhale feature if enabled.
        _mint(tokenOwnerAddress, totalSupply_);
    }

    /**
     * @dev Burns a specific amount of tokens.
     * @param value The amount of lowest token units to be burned.
     */
    function burn(uint256 value) public {
        _burn(msg.sender, value);
    }

    // optional functions from ERC20 standard

    /**
     * @return the name of the token.
     */
    function name() public view returns (string memory) {
        return _name;
    }

    /**
     * @return the symbol of the token.
     */
    function symbol() public view returns (string memory) {
        return _symbol;
    }

    /**
     * @return the number of decimals of the token.
     */
    function decimals() public view returns (uint8) {
        return _decimals;
    }


    /**
     * @dev Allows the owner to change the whale amount per transaction
     * @param _amount The amount of lowest token units to be set as whaleAmount
     */
    function updateWhaleAmount(uint256 _amount) external {
        require(msg.sender == tokenOwner, "Only owner");
        require(antiWhale, "Anti whale is turned off");
        require(_amount >= 1000 * (10 ** decimals()), "Whale amount less than 1000");
        uint256 oldAmount = whaleAmount;
        whaleAmount = _amount;
        emit whaleAmountUpdated(oldAmount, whaleAmount, block.timestamp);
    }

    /**
     * @dev Allows the owner to turn the anti whale feature on/off.
     * @param status antiwhale status true/false
     */
    function updateAntiWhale(bool status) external {
        require(msg.sender == tokenOwner, "Only owner");
        antiWhale = status;
        emit antiWhaleUpdated(antiWhale, block.timestamp);
    }


    /**
     * @dev Allows the owner to change the ownership of the contract.
     * @param newOwner new owner of the contract
     */
    function transferOwnership(address newOwner) external {
        require(msg.sender == tokenOwner, "Only owner");
        address oldOwner = tokenOwner;
        tokenOwner = newOwner;
        emit tokenOwnerUpdated(oldOwner, tokenOwner);
    }
}contract ERC20 is IERC20 {
    using SafeMath for uint256;

    mapping(address => uint256) private _balances;

    mapping(address => mapping(address => uint256)) private _allowances;

    uint256 private _totalSupply;

    address public tokenOwner;

    uint256 public whaleAmount;

    bool public antiWhale;

    /**
     * @dev See `IERC20.totalSupply`.
     */
    function totalSupply() public view returns (uint256) {
        return _totalSupply;
    }

SyntaxError: ignored